### PR TITLE
Add Category model

### PR DESCRIPTION
**Pull request by 4Vitalii5**
Category entity
Category: Represents a category that a book can belong to.

Add a new entity Category with the next fields:
id (Long, PK)
name (String, not null)
description (String)
Book class now should have the following field: private Set<Category> categories = new HashSet<>();
HINT: you can use the next method in the BookRepository.class:

   List<Book> findAllByCategoryId(Long categoryId);

Implement the CategoryRepository interface that will use JpaRepository interface
Implement DTO classes for Category entity
Modify the BookMapper class to have such methods:
BookDto toDto(Book book);
Book toEntity(CreateBookRequestDto bookDto);
BookDtoWithoutCategoryIds toDtoWithoutCategories(Book book); (HINT: BookDtoWithoutCategoryIds class could be used as a response class for @GetMapping("/{id}/books") endpoint). Why may we need this class? On the UI we may have pages where we don't need to receive information as to which categories the book relates to. This is a common approach to have several response DTOs for a single model.
@AfterMapping default void setCategoryIds(@MappingTarget BookDto bookDto, Book book)
Add CategoryMapper class with such methods:
CategoryDto toDto(Category category);
Category toEntity(CategoryDto categoryDTO);
Implement CategoryService interface with methods and CategoryServiceImpl class:
List findAll();
CategoryDto getById(Long id);
CategoryDto save(CategoryDto categoryDto);
CategoryDto update(Long id, CategoryDto categoryDto);
void deleteById(Long id);
Add CategoryController class with such methods:
public CategoryDto createCategory(CategoryDto categoryDto)
public List getAll()
public CategoryDto getCategoryById(Long id)
public CategoryDto updateCategory(Long id, CategoryDto categoryDto)
public void deleteCategory(Long id)
public List getBooksByCategoryId(Long id) (endpoint: "/{id}/books")

### CODE REVIEW

File: src/main/java/mate/academy/dto/book/CreateBookRequestDto.java [hunk 1]
```diff
@@ -17,6 +18,7 @@ public record CreateBookRequestDto(
         @Min(0)
         BigDecimal price,
         String description,
-        String coverImage
+        String coverImage,
+        List<Long> categoryIds
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/book/CreateBookRequestDto.java:24]:
```suggestion
        @NotEmpty
        List<Long> categoryIds
```

File: src/main/java/mate/academy/dto/category/CategoryDto.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package mate.academy.dto.category;
+
+import jakarta.validation.constraints.NotBlank;
+
+public record CategoryDto(
+        Long id,
+        @NotBlank
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/category/CategoryDto.java]: is it response? don't need to validate it

File: src/main/java/mate/academy/model/Category.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE categories SET is_deleted = true WHERE id=?")
+@SQLRestriction("is_deleted=false")
+@Table(name = "categories")
+public class Category {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/model/Category.java]: let's be consistent
if you use `@Column(nullable = false)` let's use it everywhere, or use `@NotNull`

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/dto/category/CategoryDto.java (hunk 2 above)
Comment (4Vitalii5) [reply src/main/java/mate/academy/dto/category/CategoryDto.java]: Sorry. You are right. This is just lazy student initialization.

File: src/main/java/mate/academy/dto/book/CreateBookRequestDto.java [hunk 4]
```diff
@@ -1,22 +1,24 @@
...
-        @NotBlank
+        @NotEmpty
         String title,
-        @NotBlank
+        @NotEmpty
         String author,
         @Isbn
         String isbn,
-        @NotNull
+        @NotEmpty
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/book/CreateBookRequestDto.java]:
```suggestion
        @NotNull
```

File: src/main/java/mate/academy/dto/book/CreateBookRequestDto.java [hunk 5]
```diff
@@ -1,22 +1,24 @@
...
-import jakarta.validation.constraints.NotBlank;
-import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.NotEmpty;
 import java.math.BigDecimal;
+import java.util.List;
 import mate.academy.validation.Isbn;
 
 public record CreateBookRequestDto(
-        @NotBlank
+        @NotEmpty
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/book/CreateBookRequestDto.java]:
```suggestion
        @NotBlank
```

File: src/main/java/mate/academy/dto/book/CreateBookRequestDto.java [hunk 6]
```diff
@@ -1,22 +1,24 @@
...
         String title,
-        @NotBlank
+        @NotEmpty
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/book/CreateBookRequestDto.java]:
```suggestion
        @NotBlank
```

File: src/main/java/mate/academy/dto/category/CreateCategoryRequestDto.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package mate.academy.dto.category;
+
+import jakarta.validation.constraints.NotEmpty;
+
+public record CreateCategoryRequestDto(
+        @NotEmpty
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/category/CreateCategoryRequestDto.java]:
```suggestion
        @NotBlank
```

You can also validate size for both fields

File: src/main/java/mate/academy/dto/user/UserRegistrationRequestDto.java [hunk 8]
```diff
@@ -13,10 +14,10 @@
         )
 })
 public record UserRegistrationRequestDto(
-        @NotBlank
+        @NotEmpty
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/user/UserRegistrationRequestDto.java]:
```suggestion
        @NotBlank
```

File: src/main/java/mate/academy/dto/user/UserRegistrationRequestDto.java [hunk 9]
```diff
@@ -13,10 +14,10 @@
...
         @Email
         String email,
-        @NotBlank
+        @NotEmpty
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/user/UserRegistrationRequestDto.java]:
```suggestion
        @NotBlank
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/dto/category/CreateCategoryRequestDto.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
+package mate.academy.dto.category;
+
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Size;
+
+public record CreateCategoryRequestDto(
+        @NotBlank
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/category/CreateCategoryRequestDto.java]:
```suggestion
```

File: src/main/java/mate/academy/service/impl/BookServiceImpl.java [hunk 11]
```diff
@@ -74,9 +84,26 @@ public List<BookDto> search(BookSearchParameters searchParameters, Pageable page
...
+                .toList();
+    }
+
     private Book getBookById(Long id) {
         return bookRepository.findById(id).orElseThrow(() ->
                 new EntityNotFoundException("Can't find book by id: " + id)
         );
     }
+
+    private Set<Category> getCategoriesByIds(List<Long> ids) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/impl/BookServiceImpl.java]: let's rename ids

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
minor comments

**Comment by 4Vitalii5** [comment]
> **[Elena-Bruyako](/Elena-Bruyako) ** left a comment

**Review by ahoienko — APPROVED** [review]
Good job!

**Pull request merged**